File: {{ cookiecutter.app_name }}/{{ cookiecutter.lib_name }}/src/main/java/{{ cookiecutter.lib_package_dir }}/APIException.java
package {{ cookiecutter.lib_package_name }};

public class APIException extends Exception {

    private final Integer codError;

    private final String message;

    public APIException(Throwable throwable) {
        super(throwable);
        this.message = "";
        this.codError = ERROR.GENERIC;
    }

    private APIException(Integer codError, Throwable throwable, String detailMessage) {
        super(detailMessage, throwable);
        this.codError = codError;
        this.message = detailMessage;
    }

    public APIException(Integer codError, String detailMessage) {
        super(detailMessage);
        this.codError = codError;
        this.message = detailMessage;
    }

    public static final class ERROR {
        public static final Integer GENERIC = 99;

        private ERROR() {
            throw new IllegalAccessError("Utility class");
        }
    }

    public APIException setCause(String message) {
        this.initCause(new Throwable(message));
        return this;
    }

    public Integer getCodError() {
        return codError;
    }


    public static APIException getFactoryException( Integer error) {
        return new APIException(ERROR.GENERIC, "Default message");
    }

    public String getAllcauses() {
        String strings = getLocalizedMessage();
        Throwable cause = this;
        while (cause.getCause() != null) {
            cause = cause.getCause();
            strings = strings.concat(" => " + cause.getLocalizedMessage());
        }
        return strings;
    }

    public String getMessage() {
        return message;
    }
}

File: {{ cookiecutter.app_name }}/{{ cookiecutter.lib_name }}/src/main/java/{{ cookiecutter.lib_package_dir }}/APIException.kt
package {{ cookiecutter.lib_package_name }} 

/**
 *
 * @author {{ cookiecutter.author_name }}
 * @version 1.0.a {% now 'local', '%d/%m/%Y' %}
 * @since 1.0.a
 */
class APIException : Exception {
    val codError: Int
    override val message: String

    constructor(throwable: Throwable?) : super(throwable) {
        message = ""
        codError = ERROR.GENERIC
    }

    private constructor(codError: Int, throwable: Throwable, detailMessage: String) : super(detailMessage, throwable) {
        this.codError = codError
        message = detailMessage
    }

    constructor(codError: Int, detailMessage: String) : super(detailMessage) {
        this.codError = codError
        message = detailMessage
    }

    class ERROR private constructor() {
        companion object {
            const val GENERIC = 99
        }

        init {
            throw IllegalAccessError("Utility class")
        }
    }

    fun setCause(message: String?): APIException {
        initCause(Throwable(message))
        return this
    }

    val allcauses: String
        get() {
            var strings = localizedMessage
            var cause: Throwable? = this
            while (cause!!.cause != null) {
                cause = cause.cause
                strings = strings + " => " + cause!!.localizedMessage
            }
            return strings
        }

    companion object {
        fun getFactoryException(error: Int?): APIException {
            return APIException(ERROR.GENERIC, "Default message")
        }
    }
}